lms/djangoapps/course_home_api/outline/v1/views.py

```diff
@@ -293,6 +293,17 @@ def get(self, request, *args, **kwargs):
 
         return Response(serializer.data)
 
+    def finalize_response(self, request, response, *args, **kwargs):
+        """
+        Return the final response, exposing the 'Date' header for computing relative time to the dates in the data.
```

Comment:
I might like to see some expansion on why we do this here. Like you say in the PR description `Browser time is notoriously unreliable for this`. Maybe add that phrase here at the least.

But also, an explanation of what you mean by that would be nice. Like, that's not a notorious fact to dumb-old me. So for me reading that, I wouldn't really have the context for why we do this weird header thing if I came across this code.

How much is browser time often off by?

This date that we send is going to be a frozen-in-time moment. If they leave the tab open for a day, is the frontend going to be able to track that and give an accurate time tomorrow? (FormattedRelative does that correctly for you I believe, but if you're overriding its logic with a custom time, I'm not sure if it can or not - but maybe you've looked at this and it does the right thing.)

Reply:
> How much is browser time often off by?

For as often as I've heard "_never_ trust the browser time", I'm having a hard time finding a good source to cite.  Anecdotally, shortly after the fall 2020 DST change, the Cybersource library we use on the checkout page started trying to validate a token that has a 15min lifetime based on browser time and 10%-20% of clients were not able to load the checkout page.

Yes, _10%-20%_ of client clocks were off by _at least_ 15mins (probably a full hour, maybe more) several days after the time-change.  I'd be interested to measure how bad the situation is more broadly, but based on that case we don't want to be showing a countdown which is that far off.

> This date that we send is going to be a frozen-in-time moment. If they leave the tab open for a day, is the frontend going to be able to track that and give an accurate time tomorrow?

Yes, the client computes a correction offset against the local clock when the response is received and uses that in relative time calculations.  Note that this correction offset will not currently update unless the API is called again even if the client clock changes significantly; for now we consider that an acceptable edge-case.  You can look at the client side usage in [my side-PR](https://github.com/edx/frontend-app-learning/pull/385) and Emma's [REV-2126 PR](https://github.com/edx/frontend-app-learning/pull/377) if you're interested.

FWIW, as far as we can tell, `FormattedRelative` does not produce a message compatible with our design spec, unfortunately.

Reply:
Ah got it, you're doing custom client code. And it makes sense for something that is a pure countdown to a server datetime to use the server's time as a seed. So all good there.

But now separately from this PR, I'm just curious about the time stuff. Your example was DST related, which would sort itself out within a week or so, I'd hope? Curious what the baseline of wrong times (before DST switch) was.

10-20% is a lot of clients that were on wrong timezone! I wonder how many devices don't auto-switch by default. I didn't think many people would turning off auto-switching and go with a manual timezone, but people with agency often surprise me.

Reply:
Yeah, I'm interested to measure this broadly, and the generalized version that computes the correction delta for every request would give us a data-source to do this; what is the distribution of client clock accuracy, are badly-inaccurate clocks more prevalent on some platforms, what timezone(s) are they set to vs. what they appear to be in, are DST time changes a factor, do they tend to get better over time after a change, spring change vs. fall change, etc.

Reply:
Note that my example was an issue that occurred shortly after a DST change, but it was a bug that was _introduced_ a couple days after the time-change, _not_ something that broke when the time-change happened.